codice/ddf
# [2.13.x] DDF-4212 Added the ability to get a CRL from an https URL

#### What does this PR do?
Adds the ability to get a CRL from an https URL

#### Who is reviewing it? 
@garrettfreibott @brjeter 

#### Select relevant component teams: 
@codice/security 

#### Ask 2 committers to review/merge the PR and tag them here.
@coyotesqrl
@ricklarsen 
@stustison

#### How should this be tested?
- Verify that the CRL works
1. Create a p12 file that can be used in Chrome using DDF's demo CA
   - `cd <DDF_HOME>/etc/certs/demoCA`
   - Create a .p12 file from the ca cert and ca key
      - `openssl pkcs12 -export -out ca.p12 -inkey private/cakey.pem -in cacert.pem`
   - Open the KeyStoreExplorer application and import the `ca.p12` file above
      - `Create a new KeyStore`
      - Select `PKCS #12`
      - Go to `Tools > Import Key Pair > PKS#12` and import the `ca.p12` file
      - Right click on the newly imported keypair and select `Sign > Sign New Key Pair`
         - Add `CN=<hostName>` to `Name`
      - Right click on the newly created keypair and select `Export > Export Key Pair`
2. Import that exported keypair into Google Chrome.
   - `Settings > Advanced > Manage Certificates`
   - On Mac:
      - `File > Import Items > Login > <Exported-KeyPair>.p12`
3. Create a CRL
   - Use the config files in [conf.zip](https://github.com/codice/ddf/files/2540088/conf.zip) (copy these files to the demoCA directory)
   - `openssl ca -config ca.conf -gencrl -keyfile private/cakey.pem -cert cacert.pem -out root.crl.pem`
   - `openssl crl -inform PEM -in root.crl.pem -outform DER -out root.crl`
4. Host CRL using python version 3
   - You will need the contents of [crl.zip](https://github.com/codice/ddf/files/2540008/crl.zip)
   - Add your crl to that file
      - Make sure it's named root.crl or edit the main.py file anc change `root.crl` to the name of your file
   - `pip3 install pyOpenSSL flask`
   - `python3 main.py`
   - The URL will be https://localhost:8443/crl_file
5. Add the CRL to DDF
   - Navigate to `System > Configuration > Certificate Revocation List (CRL)`
   - Add https://localhost:8443/crl_file under `CRL URL address`
   - `Enable CRL via URL` option
   - **Verify that a local crl is created under `<DDF_HOME>/etc/keystore`**
   - **Verify that the crl file path is added to the signature and encryption properties under `<DDF_HOME>/ws-security/issuer` and `<DDF_HOME>/ws-security/server`**
   - **Verify that the other system properties are not replaced by their actual value. For instance the keystore password is `org.apache.ws.security.crypto.merlin.keystore.password=${javax.net.ssl.keyStorePassword}` 
and not 
`org.apache.ws.security.crypto.merlin.keystore.password=changeit`**
6. Access DDF using the cert added to Chrome (may I suggest incognito mode)
   - When accessing DDF, Chrome will prompt you to choose a cert, select the one recently added
   - From DDF's login page, select `Sign in with certificate`
   - **You should be able to login**
7. Revoke the certificate 
   - Get the certificate from the p12 file. You can do 
      - `openssl pkcs12 -in destab.local.p12 -out certificate.crt` or
      - From Keystore Explorer, right click on the created keypair and select `Export > Export Certificate Chain`
   - Revoke the certificate 
      - `openssl ca -config ca.conf -revoke certificate.crt -keyfile private/cakey.pem -cert cacert.pem`
      - `openssl ca -config ca.conf -gencrl -keyfile private/cakey.pem -cert cacert.pem -out root.crl.pem`
      - `openssl crl -inform PEM -in root.crl.pem -outform DER -out root.crl`
      - If you do `openssl crl -inform DER -text -noout -in root.crl` you should see one revoked cert
      - Replace the `root.crl` file that was in the crl directory in step 4 and restart the server 
      - Re-save the configuration under `System > Configuration > Certificate Revocation List (CRL)` to make sure the CRL is updated
      - Restart DDF 
    - Try signing in with the revoked cerificate
      - Repeat step 6
      - **You should NOT be able to login**
      - **You should see a `Credentials are invalid` error**

- Update the CRL file without changing the config in DDF and verify it gets updates withn 30 minutes.
- Verify error alert
   - Enter and http url and verify that the system properties are not set and error is posted to the admin console
    - Host a non-crl file and verify that that the system properties are not set and error is posted to the admin console

#### What are the relevant tickets?
[DDF-4212](https://codice.atlassian.net/browse/DDF-4212)

#### Screenshots
Documentation:
![48224723-ef39c280-e357-11e8-8151-c84eda605398](https://user-images.githubusercontent.com/29550062/48867961-33d04f80-ed95-11e8-8ea5-c6982f5f475b.png)

Error alert:
![screen shot 2018-11-08 at 3 27 40 pm](https://user-images.githubusercontent.com/29550062/48231310-e43c5d80-e36a-11e8-9865-a4fbb0bd2309.png)

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
